pkg/upgrade/control_plane.go

```diff
@@ -279,67 +328,124 @@ func (u *ControlPlaneUpgrader) updateMachines(machines *clusterapiv1alpha1.Machi
+	// copy node registration
+	bootstrap := &bootstrapv1.KubeadmConfig{}
+	if machine.Spec.Bootstrap.ConfigRef.Namespace == "" {
+		machine.Spec.Bootstrap.ConfigRef.Namespace = u.clusterNamespace
+	}
+	err := u.managerClusterClient.Get(context.TODO(), ctrlclient.ObjectKey{Name: machine.Spec.Bootstrap.ConfigRef.Name, Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace}, bootstrap)
 	if err != nil {
 		return errors.WithStack(err)
 	}
-	if m.Annotations == nil {
-		m.Annotations = map[string]string{}
+
+	// modify bootstrap config
+	bootstrap.SetName(name)
+	bootstrap.SetResourceVersion("")
+	bootstrap.SetOwnerReferences(nil)
+
+	// find node registration
+	nodeRegistration := kubeadmv1beta1.NodeRegistrationOptions{}
+	if bootstrap.Spec.InitConfiguration != nil {
+		nodeRegistration = bootstrap.Spec.InitConfiguration.NodeRegistration
+	} else if bootstrap.Spec.JoinConfiguration != nil {
+		nodeRegistration = bootstrap.Spec.JoinConfiguration.NodeRegistration
 	}
-	m.Annotations[UpgradeIDAnnotationKey] = u.upgradeID
-	updated, err := json.Marshal(m)
+	if bootstrap.Spec.JoinConfiguration == nil {
+		bootstrap.Spec.JoinConfiguration = &kubeadmv1beta1.JoinConfiguration{
+			ControlPlane: &kubeadmv1beta1.JoinControlPlane{},
+		}
+	}
+	bootstrap.Spec.JoinConfiguration.NodeRegistration = nodeRegistration
+
+	// clear init configuration
+	bootstrap.Spec.InitConfiguration = nil
+
+	err = u.managerClusterClient.Create(context.TODO(), bootstrap)
 	if err != nil {
 		return errors.WithStack(err)
 	}
-	patch, err := jsonpatch.CreateMergePatch(original, updated)
+
+	// update machine's bootstrap reference
+	machine.Spec.Bootstrap = clusterv1.Bootstrap{
+		ConfigRef: &v1.ObjectReference{
+			Kind:       "KubeadmConfig",
+			APIVersion: bootstrapv1.GroupVersion.String(),
+			Name:       name,
+			Namespace:  bootstrap.Namespace,
+		},
+	}
+
+	return nil
+}
+
+func (u *ControlPlaneUpgrader) updateInfrastructureReference(name string, machine *clusterv1.Machine) error {
+	// copy infrastructure object
+	if machine.Spec.InfrastructureRef.Namespace == "" {
+		machine.Spec.InfrastructureRef.Namespace = u.clusterNamespace
+	}
+	infraRef, err := external.Get(u.managerClusterClient, &machine.Spec.InfrastructureRef, machine.Namespace)
+	if err != nil {
+		return err
+	}
+
+	infraRef.SetResourceVersion("")
+	infraRef.SetName(name)
+	infraRef.SetOwnerReferences(nil)
+	unstructured.RemoveNestedField(infraRef.UnstructuredContent(), "spec", "providerID")
+
+	machine.Spec.InfrastructureRef.ResourceVersion = ""
+	machine.Spec.InfrastructureRef.Name = name
+
+	// create infrastructure object
+	err = u.managerClusterClient.Create(context.TODO(), infraRef)
 	if err != nil {
 		return errors.WithStack(err)
 	}
-	if _, err := u.managementClusterAPIClient.Machines(m.Namespace).Patch(m.Name, types.MergePatchType, patch); err != nil {
+
+	return nil
+}
+
+func (u *ControlPlaneUpgrader) applyAnnotation(m *clusterv1.Machine) error {
+	patch := ctrlclient.MergeFrom(m.DeepCopy())
```

Comment:
I'd suggest using the cluster api patch helper wherever we're issuing patches. It's available here https://godoc.org/github.com/kubernetes-sigs/cluster-api/util/patch

Reply:
Tracking here: https://github.com/vmware/cluster-api-upgrade-tool/issues/83